docs: add 'env subst' command (#7959)
circleci/circleci-docs: jekyll/_cci2/env-vars.md

The `$BASH_ENV` workaround only works with `bash`, and has not been confirmed to work with other shells.
{: class="alert alert-info"}

> ### Environment Variable Substitution

Author: ```suggestion
### Environment variable substitution
```

{: #environment-variable-substitution }

The CircleCI CLI offers a wrapper around the [`envsubst`](https://github.com/a8m/envsubst) tool, both locally and available in all jobs running on CircleCI. `envsubst` is a command-line utility used to replace environment variables in text strings.

> CLI command: 

Author: ```suggestion
CLI command:
```

```sh
circleci env subst
```

> #### Usage

Author: ```suggestion
#### Usage
{: #env-subst-usage }
```

The `$BASH_ENV` workaround only works with `bash`, and has not been confirmed to work with other shells.
{: class="alert alert-info"}

### Environment variable substitution
{: #environment-variable-substitution }

The CircleCI CLI offers a wrapper around the [`envsubst`](https://github.com/a8m/envsubst) tool, both locally and available in all jobs running on CircleCI. `envsubst` is a command-line utility used to replace environment variables in text strings.

CLI command:

```sh
circleci env subst
```

#### Usage
{: #env-subst-usage }

> Within your repository create a file such as `template.json`, with value replaced by environment variable strings

Reviewer: Does the command require JSON specifically or does it accept certain other formats?

Author: Any text, it does not need to be a file of any type.

```json
{
  "foo": "$FOO",

>   "provider": "${PROVIDER}"

Reviewer: Is there a difference between using vs not using curly braces?

Author: No, this was just to show that it could handle both.

}
```

> In the CircleCI app, create the corrisponding environment variables either in the project or as a [context]({{site.baseurl/contexts}}). In this config example we will show the environment variables as if they were defined directly in the config.

Reviewer: ```suggestion
The config example below shows the corresponding environment variables as if they were defined directly within a step in the config. However, we strongly recommend creating, creating the environment variables in the CircleCI app, either in [Project Settings](/docs/set-environment-variable/#set-an-environment-variable-in-a-project/) or as a [context](/docs/contexts). 
```

```yaml
version: 2.1
jobs:
  process-template:
    docker:
      - image: cimg/base:current
    steps:
      - checkout
      - run: 
          name: Process template file
          environment:
            # Environment variables would typically be served via a context
            FOO: bar
            PROVIDER: circleci
          command: |
            circleci env subst < template.json > deploy.json
            cat deploy.json
workflows:
  env-subst-workflow:
    jobs:
      - process-template

> ```

Reviewer: Is there any chance we could include a quick explanation of what `command` does here, maybe more specifically how the operators work? Either as a comment in the snippet, or outside/after the snippet.

Output:

```sh
{
  "foo": "bar",
  "provider": "circleci"
}
```

> The `circleci env subst` command can accept text input from `stdin` or as an argument.

Reviewer: I think we could probably move this sentence higher up, perhaps before the "Usage" subheading

> 

Reviewer: I'd also suggest adding a little paragraph here for more info about the CLI, e.g.

```
For instructions on installing the CircleCI CLI locally, read the [Installing the CircleCI local CLI](docs/local-cli/) guide.
```

The config example below shows the corresponding environment variables as if they were defined directly within a step in the config. However, we strongly recommend creating, creating the environment variables in the CircleCI app, either in [Project Settings](/docs/set-environment-variable/#set-an-environment-variable-in-a-project/) or as a [context](/docs/contexts).

```yaml
version: 2.1
jobs:
  process-template:
    docker:
      - image: cimg/base:current
    steps:
      - checkout
      - run: 
          name: Process template file
          environment:
            # Environment variables would typically be served via a context
            FOO: bar
            PROVIDER: circleci
          command: |
            circleci env subst < template.json > deploy.json
            cat deploy.json
workflows:
  env-subst-workflow:
    jobs:
      - process-template
```

In this example, the `<` symbol is used to redirect the contents of the `template.json` file as _input_ to the `env subst` command, while the `>` symbol is used to redirect the output of the env subst command to the `deploy.json`.

> We could alternatively pass input to the `circleci env subst` command as an argument: `circleci env subst "hello \$WORLD"`

Reviewer: ```suggestion
You could alternatively pass input to the `circleci env subst` command as an argument: `circleci env subst "hello \$WORLD"`
```
